HTML API: Add CSS selector support

sirreal opened the pull request:
**This is not ready for final review but is ready for early feedback, _especially_ around the open questions listed below.**

Introduce CSS selector based traversal of HTML documents in the HTML API. Add new `select_all` and `select` methods to the Tag Processor and HTML Processor.

```php
// With select_all to traverse a document stopping on matching selectors
$processor = WP_HTML_Processor::create_full_parser( '<p match><div att match><em><i match></i><a match>' );
foreach ( $processor->select_all( 'p, [att], em > *' ) as $_ ) {
	assert( $processor->get_attribute( 'match' ) );
}

// With select to move to a matching selector
$processor = WP_HTML_Processor::create_full_parser( '<p match><div att match><em><i match></i><a match>' );
assert( $processor->select( 'p, [att], em > *' ) );
assert( $processor->get_attribute( 'match' ) );
assert( 'P' === $processor->get_tag() );
```

A subset of the CSS selector grammar is available as described here:

https://github.com/WordPress/wordpress-develop/blob/355c9a24e0d983813ae73e8cacc59287833d2846/src/wp-includes/html-api/class-wp-css-compound-selector-list.php#L24-L39

Notable variations from selectors specification:

**Pseudo-element selectors are not supported.** Pseudo elements will not exist in the HTML and it's unclear what benefit they would bring.

**Pseudo-class selectors are not supported.** Pseudo classes could be useful, but the logic to parse and match pseudo-class selectors would add significant complexity. There's also a lot of variety in pseudo-selectors, and rather than supporting simpler selectors (e.g. `:empty`) and not supporting more complex selectors (e.g. `:nth-child()`), pseudo-class selectors are completely unsupported. This is a clear and simple rule that greatly simplifies the implementation.

**Complex selectors have limited support.** Complex selectors are combined selectors using one of the combinators: ` ` (whitespace), `>`, `+`, or `~`. The _Tag Processor does not support complex selectors_ at all, it has no concept of document structure, and all complex selectors are structural. The _HTML Processor does not support the sibling combinators_ `+` or `~`, it only supports a parent/child or ancestor/descendant relationship. These selectors can be handled without tracking additional state in the document by analyzing breadcrumbs. Finally, only type selectors are allowed in non-final position, again because this allows matching against breadcrumbs without tracking additional state:

- Supported: `body heading > h1.page-title[attribute]`
- Unsupported (class / id selector in non final position): `#page > main`, `.page main`
- Unsupported (sibling selectors not supported): `ul li ~ li`, `ul li + li`.

Importantly, the selectors supported by the HTML Processor should be sufficient to support all core block attribute selectors according to [this PR](https://github.com/WordPress/wordpress-develop/pull/6388):

https://github.com/WordPress/wordpress-develop/blob/88e7d30892288caf35cf33b5f99faa75fa7e6e30/src/wp-includes/html-api/class-wp-html-attribute-sourcer.php#L18-L49

_Most_ of the listed selectors are also supported by the Tag Processor with the exception of the complex selectors:

```
figure > a
figure a
figure img
figure video,figure img
tbody tr
tfoot tr
thead tr
```

---

## Open questions

### Implementation

The implementation introduces a number of classes. The classes roughly correspond to different parts of the selector grammar. Parsing is handled in the `_list` classes that represent the top of the grammar. Matching logic is handled by each selector class. The selector classes implement a `matches` interface.

~I'm not happy with how the implementation is distributed. I'd like to move in one of two directions:~

Move the parsing logic for selectors into the selector classes, so each selector class is responsible for parsing itself.

**I've implemented this change.**

~OR~

~Move matching logic into the top level `*_list` classes to that the lower level selectors are only data containers.~

~My preference at this time (and the original implementation) would be the former.~

### Selector traversal APIs
- `select_all` is implemented as a generator and (except `_doing_it_wrong`) has no way to differentiate between "nothing matched" and "the selector was invalid or unsupported". It simply yields nothing in both cases.
- `select` uses `select_all` internally and has the same limitation.
- `select_all` expects the document position to remain the same in order to visit all matching tags. Because none of the supported selectors rely on stateful logic, this is not an issue at this time.

---

Todo:
- [ ] Get and address feedback on open questions.
- [ ] Split into smaller PRs for review. Specifically, the selector lists can be split into sepearate PRs starting with compound and following with complex selectors.


https://github.com/sirreal/html-api-debugger/pull/5 can be used for testing the parsing and matching behavior.

Trac ticket: https://core.trac.wordpress.org/ticket/62653


---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

sirreal committed:
WIP class skeleton

sirreal committed:
Document class

sirreal committed:
Do not support namespaced selectors

sirreal committed:
Flesh out stuff

sirreal committed:
Starting to actually parse

sirreal committed:
Add ident tests

sirreal committed:
Fix ident non-ascii bug

sirreal committed:
Use class after defined

sirreal committed:
Fix some char stuff

sirreal committed:
Improve tests

sirreal committed:
Housekeeping

sirreal committed:
Require new file in WP

sirreal committed:
Fix offset type

sirreal committed:
Add more tests and invalid tests

sirreal committed:
Fix wrong offset var usage

sirreal committed:
comment tweak

sirreal committed:
Implement codepoint escape with strspn

sirreal committed:
Test with UPPER HEX

sirreal committed:
Add ID tests

sirreal committed:
Improve tests

sirreal committed:
Add class selector tests

sirreal committed:
Add class selector

sirreal committed:
Simplify id selector parse

sirreal committed:
Improve ident tests

sirreal committed:
Add type selector tests

sirreal committed:
Add docs and remove unreachable line

sirreal committed:
Add type selector class

sirreal committed:
Add attribute selector tests

sirreal committed:
improve attr tests

sirreal committed:
Fix expectation argument order

sirreal committed:
Add test and fix is_ident

sirreal committed:
Add parse_string stub

sirreal committed:
Add attribute selector parsing

sirreal committed:
Fix test expectations

sirreal committed:
More and improved attribute tests

sirreal committed:
Implement parse_string

sirreal committed:
Add string parse tests

sirreal committed:
Remove covers annotations

sirreal committed:
Remove unused line

sirreal committed:
Improve tests for 100% coverage on parse methods

sirreal committed:
Improve documentation

sirreal committed:
Fix parse return type and return annotations

sirreal commented on src/wp-includes/html-api/class-wp-css-selectors.php:
```diff
@@ -0,0 +1,820 @@
+ * - ID selectors (e.g. `#unique-id`)
+ * - Attribute selectors (e.g. `[attribute-name]` or `[attribute-name="value"]`)
+ * - Comma-separated selector lists (e.g. `.selector-1, .selector-2`)
+ * - The following combinators:
+ *   - descendant (e.g. `.parent .descendant`)
+ *   - child (`.parent > .child`)
```
I don't think this will be supported initially. Maybe no combinators at all, only simple selectors.

These two combinators can probably be supported by the HTML API as long as they non-final selector is an element selector:

- `div > .className` ✅ supported
- `.className > div` ⛔️ not supported

Tags can be "seen" via bookmarks, while things like IDs, classes, attributes etc. would require seeking or more advanced internal tracking.

adamziel replied:
Is the problem that we'd need to keep track of all the attributes requested by the selector in the entire breadcrumbs trail?

sirreal replied:
Exactly. Either breadcrumbs would need to store all of their attributes or we'd need to seek around the document to check them which would be very costly.

All of the supported selectors are things that can be checked from given tag in the document without any seeking as the HTML API is implemented now.

adamziel replied:
If we knew the selector upfront, we could always store the attributes needed to match it, e.g. store the `class` attribute if we know we'll only ever look for `div > .className`. Only that wouldn't be practical.

However, keeping track of classnames in all the breadcrumbs elements seems okay-ish resource-wise.

Perhaps we could even have a "CSS-enabled" mode where we'd keep track of all the breadcrumbs attributes shorter than, say, 100 bytes? It wouldn't be perfect but should suffice for most attribute-based matches.

sirreal replied:
There will be ways to support more selectors, one good idea is to set some state to track data for certain selectors as the document is traversed.

One tricky thing with that implementation is that it may be necessary to return to the start of the document depending on the selectors used and scan again from there so that the necessary data is collected during traversal. This probably involves analyzing the selectors, deciding what data needs to be stored, seeking to the beginning if necessary, seeking back to the current location, and then trying to find the selector.

The complexity here is significant. It can be overcome, but I think the set of selectors supported in this PR as it is now is a good starting place with very limited matching complexity. Iteration to improve support can be added in future PRs.

adamziel replied:
Oh! Everything I said I meant for the future, no need to hold up this work :) 

On seeking - sounds reasonable, although it won't combine well with streaming. What would an example of a selector that requires backtracking?

sirreal replied:
> What would an example of a selector that requires backtracking?

This is mostly necessary to support selecting from arbitrary positions in the document. If we knew we were at the start of the document, we could determine what additional state needed to be tracked and start collecting it.

- The sibling combinators `+` and `~` select based on preceding elements, so it would be necessary to back up to check for matches.
- Support for subclass selectors (ID, class, and attribute) in non-final position, like `#my-id div` would require backing up to parse some attributes along with breadcrumbs.
- If pseudo-class selectors were added, I'd expect many of them to require more understanding of the document structure.

dmsnell replied:
in earlier work I did I actually proposed a list-interface for selecting elements and that list took all selectors of interest. that’s not a great solution to mandate, but it does allow for up-front analysis of the set of selectors, and then a crawl to find each one in succession.

I was going to make a comment about `select_all()` and the generator it yields and how I worry it miscommunicates the ability to chain selection at that point. if we start selecting a different selector from within the `foreach`, we move the underlying HTML Processor and now it will select a different element on the next iteration than it would have (potentially) if we hadn’t called one from inside.

to that end, if the underlying HTML Processor is moving, can we not get `select_all()` behavior simply by calling `select()` in succession in `while ( $processor->select() )`?

then, one thing is more explicit, that if we call `select()` from inside that loop, we kind of acknowledge we might change the outer loop because the `$processor` is shared and stateful.

it doesn’t solve the problem of the unknown necessary context from above.

thinking that the `FirTree` representation _might_ be able to help us by tracking attribute spans, making it faster to re-scan parent nodes, but also we could say someone can register selectors before traversal, and we maintain the relevant attributes as we go, regardless of which `select()` is running…

sirreal committed:
Update documentation links and grammar

sirreal committed:
Update documentation and class name

sirreal committed:
Add selector class

sirreal committed:
Implement complex selector

sirreal committed:
Working and tested

sirreal committed:
Add select method

sirreal committed:
Selector parsing should allow cap I,S modifier

sirreal committed:
CSS Add matches to selector classes

sirreal committed:
Match is successful on _any_ match in selector list

sirreal committed:
ID matches depend on quirks mode

sirreal committed:
PICKME: Add is_quirks_mode method to processor

sirreal committed:
has_class may return null, coerce to bool

sirreal committed:
Update docs to only allow subclass selectors in final complex selector position

sirreal committed:
Restrict complex selectors to only allow subclass selectors in final position

sirreal committed:
Work on complex selector handling

sirreal committed:
Implement descendent selector matching

sirreal committed:
Add null check for subclass selectors

sirreal committed:
CSS selector reformat ternaries

sirreal committed:
Implement ~= attribute matching

sirreal committed:
CSS fix return type

sirreal committed:
Fix static analysis problems

sirreal committed:
Fix and annotate things (static analysis)

sirreal committed:
update tests

sirreal committed:
Id attribute must be a string to match id selector

sirreal committed:
Coerce boolean attributes to ""

sirreal committed:
Fix a few more static analysis things

sirreal committed:
Unify parsing under single class

sirreal committed:
Rename files to align with class name

sirreal committed:
Add html processor select test suite

sirreal committed:
Fix select types

sirreal committed:
Update class doc

sirreal committed:
Improve select_ method arguments, docs, implementation

sirreal committed:
Split classes into their own files

Satisfy the 1-class-per-file requirement

sirreal committed:
Remove redundant see phpdoc annotations

sirreal committed:
Fix docs and return type on select_all

sirreal committed:
Improve html select test docs

sirreal committed:
Add select support to tag processor

Split up main CSS selector class and support more restricted selectors in the tag processor.

sirreal committed:
Simplify whitspace splitting function

sirreal committed:
Remove unreachable code

sirreal committed:
Add a lot of selector integration tests

sirreal committed:
Extract normalize input method

sirreal committed:
tests

sirreal committed:
Add nonfinal subclass selector test

sirreal committed:
Fix logic bug in child selector exploration

sirreal committed:
Improve selector integration tests

sirreal committed:
Try abstract class instead of interface

sirreal committed:
Revert "Try abstract class instead of interface"

This reverts commit 74881651faf991eabceb090707ce8b43c2a25316.

sirreal committed:
Clean up and document attribute selector

sirreal committed:
Update ticket number in tests

sirreal committed:
Improve some types

sirreal committed:
Fix and improve string token parsing

sirreal committed:
Update attribute selector tests

sirreal committed:
Revert "Update attribute selector tests"

This reverts commit 7df9ed91a1360d80c1dcb87980af941010b926ba.

sirreal committed:
Improve some complex selector match tests

sirreal committed:
Add and use matches_tag type selector method

sirreal committed:
Improve complex selector structure

Separate the self selector from relative selectors

sirreal committed:
Rework structure of complex_selector class

sirreal committed:
Improve documentation

sirreal committed:
Document complex selector class

sirreal committed:
Document matches functions

sirreal committed:
Simplify condition in compound::matches

sirreal committed:
Change class require order

sirreal committed:
Annotate matches processor argument type

sirreal committed:
Document class selector and update class_name property

sirreal committed:
Document ID selector class, rename id property

sirreal committed:
Document type selector class and rename type property

sirreal committed:
Document compound selector

sirreal committed:
Improve attribute selector docs and types

sirreal committed:
Update matches docs

sirreal committed:
Document complex selector class

sirreal committed:
Merge branch 'trunk' into html-api/add-css-selector-parser

sirreal committed:
PHP < 7.4 does not like this annotation

sirreal committed:
Update since annotations to 6.8.0

sirreal committed:
Update attr-modifier to match selectors grammar

sirreal committed:
Merge branch 'trunk' into html-api/add-css-selector-parser

sirreal commented:
[This can be tested/demoed using the playground + HTML API debugger plugin](https://playground.wordpress.net/?plugin=html-api-debugger&url=%2Fwp-admin%2Fadmin.php%3Fpage%3Dhtml-api-debugger%26html%3D%253Cmeta%2Bproperty%253D%2522og%253Anice%2522%2Bcontent%253D%2522that%2527s%2Bright%2522%253E%250A%253Cfoo%253E%250A%253Cbar%2Batt-bar%253E%250A%253Cbaz%2Bclass%253D%2522foo%2Bbar%2Bbaz%2522%253E%250A%253Cquux%2Bid%253D%2522wonky%2522%253E%26selector%3Dfoo%252C%250A%255Batt-bar%255D%252C%250A.baz%252C%250A%2523wonky%252C%250Ameta%255Bproperty%255E%253D%2522OG%253A%2522i%255D&core-pr=7857)

sirreal commented on src/wp-includes/html-api/class-wp-css-complex-selector.php:
```diff
@@ -0,0 +1,192 @@
+	 * @param string[] $breadcrumbs Breadcrumbs.
+	 * @return bool True if a match is found, otherwise false.
+	 */
+	private function explore_matches( array $selectors, array $breadcrumbs ): bool {
```
This is a recursive function. It's bounded by the length of the breadcrumbs list and the depth of the complex selector, but it may be good to set some limits here.

sirreal commented:
I'm not sure how useful it would be, but I think it would be interesting to add namespaces to type selectors. We could declare 4 namespaces to support: `*`, `html`, `svg`, and `math`. Then things could be selected like:

- Any MathML element: `math|*`
- SVG Title elements: `svg|title`

sirreal committed:
Merge branch 'trunk' into html-api/add-css-selector-parser

sirreal committed:
Merge branch 'trunk' into html-api/add-css-selector-parser

sirreal committed:
Move parsing back to selector classes

sirreal committed:
Update tests for class parsing

sirreal committed:
Use whitepsace chars constant

sirreal committed:
parse_whitespace should be protected

sirreal committed:
Update interface to abstract class require

sirreal committed:
Document base class

sirreal committed:
Invert and comment confusing compound selector condition

sirreal committed:
Use switch in compound selector parsing

sirreal committed:
Fix up some todo-s

sirreal committed:
Make most selector constructors private

sirreal commented:
> I'm not happy with how the implementation is distributed. I'd like to move in one of two directions:
>
> Move the parsing logic for selectors into the selector classes, so each selector class is responsible for parsing itself.

I've changed the class structures so that all the selectors extend an abstract base class that provides some constants and utilities for parsing selectors. Each selector class now defines how it is parsed to construct an instance of itself.

Most of the classes expose properties that should be private and only used in their internal match methods. The properties are inspected in tests now, but this should be changed.

sirreal committed:
Fix test class implementation of abstract class

sirreal committed:
Remove php 8+ ?static return types

sirreal committed:
Fix typo in Exception class name

sirreal committed:
Remove ?static return type from test

sirreal committed:
Merge branch 'trunk' into html-api/add-css-selector-parser

dmsnell commented on src/wp-includes/html-api/class-wp-css-attribute-selector.php:
```diff
@@ -0,0 +1,367 @@
+	 *
+	 *     [attr|=value]
+	 */
+	const MATCH_EXACT_OR_HYPHEN_PREFIXED = 'exact-or-hyphen-prefixed';
```
this reads like the hyphen comes first, but in the CSS selector, it specifically connotes that a hyphen _follows_ the match. would `HYPHEN_SUFFIXED` be more accurate?

sirreal replied:
It's difficult to make a brief name the describes this. I agree suffix is better than prefix.

016d897f35d7b59400ba4b2fd386065b187e8e4d

dmsnell commented on src/wp-includes/html-api/class-wp-css-attribute-selector.php:
```diff
@@ -0,0 +1,367 @@
+
+		if ( null === $this->value ) {
+			return true;
+		}
```
here it reads as if `null` is a sentinel value for “as long as the attribute exists” but without reading the code I would have assumed passing `null` would mean “only match if the attribute _doesn’t exist_ on the tag”

what do you think about this?

sirreal replied:
Attribute selectors have the form `[attr]` (presence) or one of the matching forms like `[att=val]` (`att` attribute is an _exact match_ for the value `val`).

The `[attr]` form is represented here with a `null` matcher (there's no matching logic) as well as a `null` value (there's no value to match). Those must either both be `null` or there must be both a matcher and a value.

I believe this is interchangeable here in case it's clearer:

```php
if ( null === $this->matcher ) {
	return true;
}
```

Another option would be to add another matcher value like `MATCH_PRESENCE` to indicate that type of matcher.

Either way, the `null` value seems appropriate to me. The selector does not have a value to match. There is no negative attribute selection.

This may be a question of perspective, for example the HTML API returns `null` if a an unset attribute is requested, `true` for a boolean attribute, or a `string` for an attribute with a value. Attribute selectors don't follow the same logic. There's no way to only match a boolean or to match elements without an attribute, so  I don't think we should try to align on those types of values. There may or may not be a string value to match.

This should mostly be an internal detail, it's not something users should really be interacting with.

I'm open to discuss this further. Either way, it's likely a good idea to clarify the `null` value in this property's documentation block.

dmsnell commented on src/wp-includes/html-api/class-wp-css-attribute-selector.php:
```diff
@@ -0,0 +1,367 @@
+	 * @return bool True if the processor's current position matches the selector.
+	 */
+	public function matches( WP_HTML_Tag_Processor $processor ): bool {
+		$att_value = $processor->get_attribute( $this->name );
```
not a critical point, but we have largely used `attribute` or `attr` in the HTML API and in Gutenberg code when dealing with attributes. Do we need to save bits here?

sirreal replied:
Attr is better, switched in c0fa8d54a4a396ddf25b12aeb51462302b33e49c.

dmsnell commented on src/wp-includes/html-api/class-wp-css-attribute-selector.php:
```diff
@@ -0,0 +1,367 @@
+						return true;
+					}
+				}
+				return false;
```
while this seems fine for now, I suspect that at some point we will prefer to crawl through the attribute value comparing as we go with `substr_compare()` rather than allocate each substring. we can remember that in some documents we see DoS attacks due to long attribute values.

`$this->whitespace_delimited_list()` could also return a starting byte offset into the value instead of the substring, which would resolve this without changing much of the calling interface.

dmsnell commented on src/wp-includes/html-api/class-wp-css-attribute-selector.php:
```diff
@@ -0,0 +1,367 @@
+				}
+
+				$starts_with = "{$this->value}-";
+				return 0 === substr_compare( $att_value, $starts_with, 0, strlen( $starts_with ), $case_insensitive );
```
seems like this whole thing could be collapsed into a single call to `substr_compare()` with a final test of the following character…

```php
$exact_length   = strlen( $this->value );
$matches_prefix = substr_compare( $att_value, $this->value, 0, $exact_length, $case_insensitive );
return (
	0 === $matches_prefix &&
	( strlen( $att_value ) === $exact_length || '-' === $att_value[ $exact_length ] )
);
```

sirreal replied:
That does seem to be correct and simpler, nice 👍 cfa2bc2ecb17d2893cb239db168608f295ee187f

dmsnell commented on src/wp-includes/html-api/class-wp-css-attribute-selector.php:
```diff
@@ -0,0 +1,367 @@
+	 *                       will be updated if the parse is successful.
+	 * @return static|null The selector instance, or null if the parse was unsuccessful.
+	 */
+	public static function parse( string $input, int &$offset ) {
```
not immediately a fan of mutating the offset passed into the function. did you consider some other examples of passing in something like `&$bytes_parsed`? for example, the HTML decoder does this which leaves the input variables untouched while still communicating `&$match_byte_length`

dmsnell commented on src/wp-includes/html-api/class-wp-css-attribute-selector.php:
```diff
@@ -0,0 +1,367 @@
+		}
+		++$updated_offset;
+
+		self::parse_whitespace( $input, $updated_offset );
```
see above: would be nice if instead of `parse_whitespace` we had this function return the length of the whitespace bytes and then this line could use the same variable names as the HTML API with

```php
$at += self::skip_whitespace( $input, $at );
```

dmsnell commented on src/wp-includes/html-api/class-wp-css-attribute-selector.php:
```diff
@@ -0,0 +1,367 @@
+					$attr_matcher    = WP_CSS_Attribute_Selector::MATCH_SUFFIXED_BY;
+					$updated_offset += 2;
+					break;
+				case '*':
```
when I explored this long ago, I actually felt like the symbols in use in CSS provided reasonable literal values in the code vs. the use of `const`s. it didn’t have the same explanatory power in the code, but it provided a more direct correspondence with the CSS, which was nice, plus involved fewer translations/CPU cycles.

sirreal replied:
I'm open to that, what do you have in mind, a single character?

- `=`: exact
- `*`: contains
- etc.

dmsnell commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -637,6 +637,75 @@ public function get_unsupported_exception() {
+	 * @param string $selector_string Selector string.
+	 * @return Generator<void> A generator pausing on each tag matching the selector.
+	 */
+	public function select_all( $selector_string ): Generator {
```
I’m not sure this is necessary because the interface of `select()` would provide the same interface as this, but without the dummy `$_` variable, would it not?

```php

while ( $processor->select( 'meta[property^="og:" i]' ) ) {
	…
}
```

except here we also have a generator and non-true/false return values. is there a reason to add a second almost-identical interface to the main loop?

I ask because I think it’s important that developers not have too many ways to do the same thing, and we’ve already tried to communicate the main interface to the HTML API as being this new "loop" in a `while ( $processor->next_token() )` or `while ( $processor->next_tag() )` construct.

sirreal replied:
I agree with this. We addressed it in dc789e4.

dmsnell reviewed (commented):
some leftover thoughts we can get back to later, or in person

sirreal committed:
Merge branch 'trunk' into html-api/add-css-selector-parser

sirreal committed:
Change MATCH_EXACT_OR_HYPHEN_PREFIXED to _SUFFIXED

This is a more appropriate name for the type of match.

> `[att|=val]` Represents an element with the att attribute, its value
> either being exactly "val" or beginning with "val" immediately
> followed by "-" (U+002D). This is primarily intended to allow language
> subcode matches (e.g., the hreflang attribute on the a element in
> HTML) as described in BCP 47 ([BCP47]) or its successor.

sirreal committed:
Use "attr" instead of "att" as short "attributes"

sirreal committed:
Simplify exact or hyphen suffixed implementation

dmsnell committed:
Only expose `select()` for a `while ( $->select() )` loop

Matches the calling interface for the other HTML API classes, avoids
creating the `Generator`, using a static var to avoid re-parsing the
selector string instead.

sirreal committed:
Fix expectedIncorrectUsage method name

sirreal committed:
Improve and fix complex selector list documentation

sirreal committed:
Add unsupports sibling selector tests

sirreal committed:
Do not support + and ~ selectors

sirreal committed:
Update @since tags to WP_VERSION placeholder

sirreal committed:
Add unsupported complex selector test

sirreal committed:
Fix spelling and grammar in documentation

sirreal committed:
Improve documentation

sirreal committed:
More documentation improvements

sirreal committed:
Avoid redundant get_attribute call

sirreal committed:
Reformat some documentation

dmsnell committed:
Rename files and remove duplicate character

sirreal committed:
Merge branch 'trunk' into html-api/add-css-selector-parser

sirreal committed:
CSS selector: Fix off-by-one rejecting [a=b] at end of selector

The length guard before the attribute matcher required 4 remaining bytes
where the minimum valid tail `=x]` is 3, so a valid exact-match attribute
selector with a single-character unquoted value at the end of the selector
string (e.g. `[a=b]`) was wrongly rejected as unparseable.

Relax the guard from `>=` to `>`. All reads after the guard are bounded:
the operator reads touch at most offset+1, and every later read re-checks
the length itself.

Adds the exact-fit valid case and invalid cases at the same boundary
(`[a=]`, `[a~=]`, `[a==b]`, `[a=1]`) to the parse tests, plus an
assertNotNull so parse failures report cleanly instead of erroring on a
null property read.

Found by the CSS selector fuzzer (tools/css-selector-fuzz, Bug 3 in
FINDINGS.md).

(cherry picked from commit 16d03e2c5fd49d79f413562d88d7e6e36774618a)

sirreal committed:
CSS selector: Empty-operand substring attribute matchers match nothing

Per Selectors level 4, the substring attribute matchers with an empty
value — [x^=""], [x$=""], [x*=""] — represent nothing and must never
match. The matcher instead matched any element carrying the attribute
(prefix and contains) or an element whose attribute value was exactly
empty (suffix).

Add an early return for the empty operand on those three matchers,
before the case modifier and boolean-attribute normalization so
[x^="" i] and valueless attributes are covered too.

[x=""], [x|=""], and [x~=""] are unaffected and remain spec-correct:
exact and hyphen matchers may match an empty value, and the one-of
matcher already matched nothing because a whitespace-delimited list
never yields an empty item. Tests pin all of these, including |= against
a hyphen-prefixed value.

https://www.w3.org/TR/selectors-4/#attribute-substrings

Found by the CSS selector fuzzer (tools/css-selector-fuzz, Bug 2 in
FINDINGS.md).

(cherry picked from commit 0cefeb2fc8fd6b255498c6dc0438b7334698b975)

sirreal committed:
CSS selector: Decode identity escapes at the byte offset, not char index

consume_escaped_codepoint() read the escaped codepoint for non-hex
(identity) escapes with mb_substr( $input, $offset, 1 ), but $offset is
a byte offset while mb_substr()'s second argument is a character index.
Any multibyte content earlier in the selector string shifts the read one
character right per continuation byte, decoding the wrong codepoint:
'Ü\sup' parsed as ident 'Üuup', and the corruption threads across an
entire selector list ('#ÜÜÜ,\sup #x' parsed the second selector's type
as ' up'). Depending on the mis-decoded codepoint this also caused
spurious parse failures of valid selectors. Hex escapes were already
byte-correct and are unaffected.

Read the codepoint from the byte offset instead. ASCII-only inputs are
byte-for-byte unchanged, and the returned codepoint's byte length keeps
the offset advancing exactly past it.

Adds parse_ident and parse_string cases pinning identity escapes after
multibyte characters, plus a hex-escape control.

Found by the CSS selector fuzzer (tools/css-selector-fuzz, Bug 1 in
FINDINGS.md).

(cherry picked from commit 7419a9fef6c6c23d14b8ae9c04ae1c91834824b0)

sirreal committed:
CSS selector: Backslash at end of input is a valid escape (U+FFFD)

Per CSS Syntax 3, a backslash followed by EOF is a valid escape in
ident context -- §4.3.8 rejects only a newline as the second code
point, and EOF is not a newline -- and consuming it returns U+FFFD
REPLACEMENT CHARACTER (§4.3.7). WP rejected the whole selector:
next_two_are_valid_escape() required a code point after the backslash,
so '.foo\' parsed to null instead of the class "foo\u{FFFD}".

Fix: consume_escaped_codepoint() returns U+FFFD at EOF without
advancing, and next_two_are_valid_escape() accepts a backslash as the
final byte. String context is unaffected: parse_string() guards EOF
itself before consuming an escape, preserving the §4.3.5 'do nothing'
EOF rule ('foo\ still parses to foo).

Review of the fix surfaced a second bug in the same family:
normalize_selector_input() trimmed *trailing* whitespace before
tokenizing, so '.foo\ ' (escaped space: the valid, unmatchable class
'foo ') and ".foo\\n" (invalid escape: must be rejected) both
collapsed to '.foo\' and matched elements with class "foo\u{FFFD}" --
a wrong-match-set bug, where before the EOF-escape fix the collapse
was a harmless fail-safe rejection. Now only leading whitespace is
stripped; the grammar already consumes insignificant trailing
whitespace via parse_whitespace() in both selector-list parsers.

Verified against lexbor: '.foo\' matches class "foo\u{FFFD}", lone
'\' parses as type U+FFFD and matches nothing, '.foo\ ' is valid and
matches nothing, and the LF/CR/FF escape variants are rejected --
exact agreement on all probes. (NEXT-STEPS.md 'candidate finding 4',
now confirmed and closed.)

(cherry picked from commit 203858bbd470a2b7a8eeaf7f7579d85eeced3264)

sirreal committed:
CSS selector: End of input auto-closes an open attribute selector

Per CSS Syntax 3 §5.4.8/§4.3.5, tokenization auto-closes unterminated
simple blocks and unterminated strings at EOF (a parse error, but the
block/string is returned), and the selector grammar then applies to the
block contents. So '[att=val' is the same selector as '[att=val]', and
'[att="a b' carries the string value 'a b'. WP rejected all of these
with null.

The attribute parser now treats the end of input like a closing ']' at
the two positions where the grammar is complete (after the name, and
after the value/modifier), and the early length guards that required
room for a closing bracket are relaxed accordingly. Truncation inside
the grammar itself is still invalid: '[', '[a=', '[a~', '[a=b x', and
a comma inside the open block ('[a=b, div') all stay null.

Escape interplay (verified per spec and in Chromium): '[a=b\' carries
the value "b\u{FFFD}" (escape at EOF in ident context), while
'[a="b\' carries 'b' (backslash-then-EOF in a string 'does nothing').
'[a\]' parses as a presence selector for the attribute 'a]' (the
escaped ']' joins the ident and EOF closes the block).

Chromium agrees with every accepted and rejected form above. lexbor
rejects all EOF-truncated forms (it does not implement the auto-close
rule) and diverges from browsers and the spec here; the fuzzer's lexbor
differential is unaffected because it compares canonical re-renders,
which always include the closing bracket.

(cherry picked from commit 5eea359bd5da8d0f8bc01a510237eec4177a5c1e)

sirreal committed:
CSS selector: Implement HTML's case-insensitive attribute value list

HTML defines 46 attributes (type, rel, lang, dir, media, hreflang,
http-equiv, ...) whose values must match ASCII case-insensitively in
attribute selectors on an HTML element when the selector carries no
i/s modifier:

https://html.spec.whatwg.org/multipage/semantics-other.html#case-sensitivity-of-selectors

WP honored only the explicit modifiers, so [type=TEXT] silently failed
to match <input type="text"> — a wrong match set rather than a
refusal, invisible to callers.

The matcher now folds case when all three hold: no modifier on the
selector, the element is in the html namespace (per the processor's
get_namespace()), and the lowercased attribute name is in the list.
An explicit s modifier still forces case-sensitive matching, per
Selectors 4 §6.3: 'the UA must match the value case-sensitively ...
regardless of document language rules.' All six matchers and |='s
hyphen check honor the rule via the existing case-insensitive
comparison branches.

Namespace scoping follows the spec's 'on an HTML element' wording:
SVG/MathML elements keep case-sensitive matching, while elements at
HTML integration points (e.g. inside <svg><foreignObject>) fold, since
they are html-namespace. Verified in Chromium, which agrees on the
integration point but also folds plain SVG-namespace elements,
diverging from the spec's scoping; WP follows the spec. The standalone
Tag Processor tracks no namespaces and folds everywhere — the same
class of approximation as its ancestor-blind matching. The review
panel machine-diffed both list constants against the live spec (exact,
in spec order).

(cherry picked from commit 40640d173ec680a05cf06bd9829273ab2a8805ae)

sirreal committed:
CSS selector: Align test data provider arrows per WPCS

phpcbf reports 40 WordPress.Arrays.MultipleStatementAlignment warnings in
this file's data providers, and the coding-standards workflow runs phpcs
over the test suite without -n, so warnings fail CI. Pure whitespace;
no test changes.

(cherry picked from commit 3db43eaef2107421f156e6ac2a29478fe6e4b2fd)

sirreal committed:
CSS selector: Decode identity escapes without copying the input tail

The identity arm of consume_escaped_codepoint() read one character via
mb_substr( substr( $input, $offset ), 0, 1 ), copying the entire
remaining input per escape: O(n^2) over selectors composed of escapes,
plus an O(n) temporary allocation each time.

Size the code point in place instead with the bounded scanner
_wp_scan_utf8( $input, $at, $invalid_length, 4, 1 ) from
compat-utf8.php (WP 6.9, loaded unconditionally before the HTML API),
then copy at most 4 bytes. Escapes of invalid UTF-8 fall through to the
literal previous mb_substr() line, so behavior is preserved by
construction under every mb_substitute_character setting; that fallback
remains O(tail) per call, accepted for developer-supplied selectors.
_wp_utf8_codepoint_span() is deliberately not used: it leaves the
scanner's ASCII fast-path unbounded, which is quadratic again (noted
in-code).

200KB of repeated \g through parse_ident: 180 ms before, 45 ms after,
with linear scaling after (47/90/180 ms at 200/400/800KB; previously
~4x per doubling) and half the peak memory. Escape pin coverage grows
to 14 cases: 2/3/4-byte characters including at end of input, NUL, and
each invalid-byte class (lone continuation, overlong lead, invalid
lead, truncated 3/4-byte, encoded surrogate, above U+10FFFF), with
expectations probe-verified against the pre-change implementation.

Adversarial review: equivalence reviewer ran ~74M differential
old-vs-new cases (exhaustive byte-class boundaries at every offset,
random fuzz, non-default mb_substitute_character) with 0 mismatches;
perf reviewer independently reproduced the quadratic-before /
linear-after curves; integration reviewer verified load order
(including SHORTINIT), private-function precedent, and phpcs. All
approved. Gates: full html-api PHPUnit group green (1654 tests),
fuzzer 5000 seeds 0 failures.

(cherry picked from commit 9d82c1ccafbf4a8db412f7873e01eb55a9b0c627)

sirreal committed:
CSS selector: Pin a canary mb_substitute_character in escape tests

Decoding an identity escape of invalid UTF-8 leaks the process-global
mb_substitute_character() setting into parse results: the substitute
character is returned and the offset advances by the byte length of the
substitute, not of the invalid sequence. Under the default '?' this is
nearly invisible; under a multibyte substitute it swallows following
characters and can push the offset past the end of the input.

Pin the setting to a distinctive canary -- U+2603 SNOWMAN -- in
set_up()/tear_down() and rewrite the seven invalid-byte pins to the
canary expectations, making the dependence unmistakable: five cases
show the trailing 'z' being eaten, and a dedicated test asserts the
offset overrun that the rest-of-input assertion cannot see (substr()
returns '' both at and past the end). A differential run of all
provider cases under canary/default/'none' confirms exactly these seven
react to the setting; everything else is independent of it.

These pins document the leak, not endorse it. They are the ready-made
red suite for the planned fix: decoding invalid bytes to U+FFFD per
maximal subpart (CSS Syntax 3 section 3.2 via the WHATWG Encoding
Standard) makes the outputs setting-independent and flips every one of
these expectations. Adversarial review approved; full html-api group
green (1654 tests) with the substitute character verified restored
after the run.

(cherry picked from commit 9b0b1df6ec47b2937e9eb49a4388b08bb0f5e104)

sirreal committed:
CSS selector: Scrub invalid UTF-8 selector input to U+FFFD

Selector strings are UTF-8 text. from_selectors() now decodes the input
byte stream before parsing: normalize_selector_input() replaces each
maximal subpart of an ill-formed byte sequence with U+FFFD via
wp_scrub_utf8() (WP 6.9), per the byte-decoding step CSS Syntax 3
section 3.2 defines through the WHATWG Encoding Standard's UTF-8
decoder. A replaced selector is almost always a developer mistake
(mojibake, double encoding) that would otherwise yield a silently empty
match set, so the replacement also reports _doing_it_wrong(), named
"<called class>::from_selectors" via late static binding.

The mb_substitute_character() leak in consume_escaped_codepoint() dies
structurally: with all public input scrubbed, the identity arm's
mb_substr() fallback became unreachable through from_selectors() and is
replaced by a deterministic decode for direct parse() callers — consume
the maximal subpart the existing _wp_scan_utf8() call already reported
and return one U+FFFD, consistent with the scrub. This also removes the
remaining O(tail)-per-escape copy for invalid bytes.

Design decision: reject (wp_is_valid_utf8() -> null) and raw byte
passthrough were both considered and discarded by a three-persona
adversarial design review; scrub is the option that stays stable under
both the current raw value getters and a future where the getters scrub
their return values. An escape-arm-only U+FFFD decode was ruled out
unanimously: it would break the identity property that escaping a
non-special code point is equivalent to writing it unescaped. The known
divergence is pinned in a test: a scrubbed selector cannot match raw
invalid bytes in a document (the Tag Processor reports raw bytes); if
the HTML API value getters are ever changed to scrub, that pin flips to
a match and must be updated in the same change. The compound-list class
docblock gains a "Text Encoding" section recording the contract.

Tests: the seven U+2603-canary escape pins flip to maximal-subpart
U+FFFD expectations, and the canary is retained permanently — its job
inverted from documenting the leak to proving setting-independence (a
reintroduced mb_substitute_character dependence fails eight tests). New
coverage: scrub + notice through from_selectors() on both list classes
(the complex-list test pins the late-static-binding notice name), a
lone invalid byte parsing as a U+FFFD type selector, the notice firing
even when the scrubbed selector is rejected by the grammar, string-token
invalid-byte decode, identity-escape equivalence and U+FFFD matching
through select(), and the raw-document-bytes no-match pin (deliberately
unique 0xC1 byte: select() memoizes the last parsed selector string, so
a unique selector guarantees the parse-time notice under any test
order).

Adversarial review: three hostile reviewers. The equivalence reviewer
verified the decode against an independent reference WHATWG UTF-8
decoder (exhaustive 1-2-byte tails, ~204k boundary-alphabet 3-4-byte
tails, 100k random; zero mismatches, all under a U+2603 canary), scrub
idempotence and ordering-neutrality (200k cases), and the notice-name
propagation. The test reviewer killed four core mutations against the
suite and demonstrated two test defects (select()-cache coupling, an
unpinned complex-list notice name), both fixed before commit. The
integration reviewer verified worker-model equivalence and determinism
(10000 fuzz seeds clean). All approved. Gates: full html-api group
green (1665 tests), fuzzer 5000 seeds 0 failures, self-check OK, phpcs
clean.

(cherry picked from commit 598ed6f363daec3d954cf22cee71b7f98f1f17c2)

sirreal committed:
Merge branch 'trunk' into html-api/add-css-selector-parser

sirreal committed:
Merge branch 'trunk' into html-api/add-css-selector-parser

sirreal committed:
HTML API: Preserve escaped type selector asterisks

sirreal committed:
Merge branch 'trunk' into html-api/add-css-selector-parser

sirreal commented on src/wp-includes/html-api/css/class-wp-css-selector-parser-matcher.php:
```diff
@@ -0,0 +1,573 @@
+	 *
+	 * This implementation is not interested in the <delim-token>, a '#' delim token is not relevant for selectors.
+	 */
+	final protected static function parse_hash_token( string $input, int &$offset ): ?string {
+		if ( $offset + 1 >= strlen( $input ) || '#' !== $input[ $offset ] ) {
+			return null;
+		}
+
+		$updated_offset = $offset + 1;
+		$result         = self::parse_ident( $input, $updated_offset );
+
+		if ( null === $result ) {
+			return null;
+		}
+
+		$offset = $updated_offset;
+		return $result;
+	}
```
This is really only parsing hash tokens if they have the ident flag I believe, because that's what's interesting for selectors.

It should be clearly noted in the method documentation.

sirreal committed:
Merge branch 'trunk' into html-api/add-css-selector-parser
